// @subql/common/dist/project/readers/github-reader.js
"use strict";
// Copyright 2020-2025 SubQuery Pte Ltd authors & contributors
// SPDX-License-Identifier: GPL-3.0
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GithubReader = void 0;
const console_1 = require("console");
const path_1 = __importDefault(require("path"));
const axios_1 = __importDefault(require("axios"));
const js_yaml_1 = __importDefault(require("js-yaml"));
const utils_1 = require("../utils");
class GithubReader {
    key;
    api;
    defaultBranch;
    constructor(key) {
        this.key = key;
        this.api = axios_1.default.create({
            baseURL: `https://raw.githubusercontent.com/${key}`,
        });
    }
    get root() {
        return undefined;
    }
    async getPkg() {
        return this.getFile('package.json');
    }
    async getProjectSchema() {
        // Github reader not support ts manifest.
        const projectYaml = await this.getFile(utils_1.DEFAULT_MANIFEST);
        return js_yaml_1.default.load(projectYaml);
    }
    async getFile(fileName) {
        try {
            const branch = await this.getDefaultBranch();
            const { data } = await this.api.get(path_1.default.join(branch, fileName));
            return data;
        }
        catch (err) {
            throw new Error('Failed to fetch file from github', { cause: err });
        }
    }
    async getDefaultBranch() {
        if (this.defaultBranch) {
            return this.defaultBranch;
        }
        const { data } = await axios_1.default.get(`https://api.github.com/repos/${this.key}`);
        (0, console_1.assert)(data.default_branch, 'Failed to fetch default branch from github');
        this.defaultBranch = data.default_branch;
        return this.defaultBranch;
    }
}
exports.GithubReader = GithubReader;
