// @subql/common/dist/project/utils.js
"use strict";
// Copyright 2020-2025 SubQuery Pte Ltd authors & contributors
// SPDX-License-Identifier: GPL-3.0
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IsEndBlockGreater = exports.tsProjectYamlPath = exports.FileReferenceImp = exports.SemverVersionValidator = exports.DEFAULT_GIT_IGNORE = exports.DEFAULT_ENV_DEVELOP_LOCAL = exports.DEFAULT_ENV_LOCAL = exports.DEFAULT_ENV_DEVELOP = exports.DEFAULT_ENV = exports.DEFAULT_TS_MANIFEST = exports.DEFAULT_MANIFEST = exports.DEFAULT_MULTICHAIN_TS_MANIFEST = exports.DEFAULT_MULTICHAIN_MANIFEST = void 0;
exports.isFileReference = isFileReference;
exports.handleCreateSubqueryProjectError = handleCreateSubqueryProjectError;
exports.makeTempDir = makeTempDir;
exports.getProjectRootAndManifest = getProjectRootAndManifest;
exports.getMultichainManifestPath = getMultichainManifestPath;
exports.validateSemver = validateSemver;
exports.delay = delay;
exports.validateObject = validateObject;
exports.extensionIsTs = extensionIsTs;
exports.extensionIsYamlOrJSON = extensionIsYamlOrJSON;
exports.forbidNonWhitelisted = forbidNonWhitelisted;
exports.IsNetworkEndpoint = IsNetworkEndpoint;
exports.normalizeNetworkEndpoints = normalizeNetworkEndpoints;
exports.notifyUpdates = notifyUpdates;
exports.toJsonObject = toJsonObject;
exports.mapToObject = mapToObject;
const fs_1 = __importDefault(require("fs"));
const os_1 = __importDefault(require("os"));
const path_1 = __importDefault(require("path"));
const class_transformer_1 = require("class-transformer");
const class_validator_1 = require("class-validator");
const yaml = __importStar(require("js-yaml"));
const semver_1 = require("semver");
const update_notifier_1 = __importDefault(require("update-notifier"));
const constants_1 = require("../constants");
exports.DEFAULT_MULTICHAIN_MANIFEST = 'subquery-multichain.yaml';
exports.DEFAULT_MULTICHAIN_TS_MANIFEST = 'subquery-multichain.ts';
exports.DEFAULT_MANIFEST = 'project.yaml';
exports.DEFAULT_TS_MANIFEST = 'project.ts';
exports.DEFAULT_ENV = '.env';
exports.DEFAULT_ENV_DEVELOP = '.env.develop';
exports.DEFAULT_ENV_LOCAL = '.env.local';
exports.DEFAULT_ENV_DEVELOP_LOCAL = '.env.develop.local';
exports.DEFAULT_GIT_IGNORE = '.gitignore';
function isFileReference(value) {
    return value?.file && typeof value.file === 'string';
}
// Input manifest here, we might need to handler other error later on
function handleCreateSubqueryProjectError(err, pjson, rawManifest, logger) {
    if (JSON.stringify(err.message).includes(constants_1.RUNNER_ERROR_REGEX)) {
        logger.error(`Failed to init project, required runner is ${rawManifest.runner.node.name}, got ${pjson.name}`);
    }
    else {
        logger.error(err, 'Create Subquery project from given path failed!');
    }
}
async function makeTempDir() {
    const sep = path_1.default.sep;
    const tmpDir = os_1.default.tmpdir();
    const tempPath = await fs_1.default.promises.mkdtemp(`${tmpDir}${sep}`);
    return tempPath;
}
function getProjectRootAndManifest(subquery) {
    const project = {
        root: '',
        manifests: [],
    };
    const stats = fs_1.default.statSync(subquery);
    if (stats.isDirectory()) {
        project.root = subquery;
        if (fs_1.default.existsSync(path_1.default.resolve(subquery, exports.DEFAULT_MANIFEST))) {
            project.manifests.push(path_1.default.resolve(subquery, exports.DEFAULT_MANIFEST));
        }
        // Then check for a 'multichain manifest'
        else if (fs_1.default.existsSync(path_1.default.resolve(subquery, exports.DEFAULT_MULTICHAIN_MANIFEST))) {
            const multichainManifestContent = yaml.load(fs_1.default.readFileSync(path_1.default.resolve(subquery, exports.DEFAULT_MULTICHAIN_MANIFEST), 'utf8'));
            if (!multichainManifestContent.projects || !Array.isArray(multichainManifestContent.projects)) {
                throw new Error('Multichain manifest does not contain a valid "projects" field');
            }
            addMultichainManifestProjects(subquery, multichainManifestContent, project);
        }
        else {
            throw new Error(`Unable to resolve manifest file from given directory: ${subquery}`);
        }
    }
    else if (stats.isFile()) {
        const { dir, ext } = path_1.default.parse(subquery);
        if (!extensionIsTs(ext) && !extensionIsYamlOrJSON(ext)) {
            throw new Error(`Extension ${ext} not supported for project ${subquery}`);
        }
        project.root = dir;
        let projectYamlPath = subquery;
        if (extensionIsTs(ext)) {
            projectYamlPath = (0, exports.tsProjectYamlPath)(subquery);
            if (!fs_1.default.existsSync(projectYamlPath)) {
                throw new Error(`Could not find manifest ${projectYamlPath}, if pointing to a typescript manifest, please ensure build successfully`);
            }
        }
        const multichainManifestContent = yaml.load(fs_1.default.readFileSync(projectYamlPath, 'utf8'));
        // The project manifest could be empty
        if (multichainManifestContent === null) {
            throw new Error(`Read manifest content is null, ${projectYamlPath}`);
        }
        else if (multichainManifestContent.projects && Array.isArray(multichainManifestContent.projects)) {
            addMultichainManifestProjects(dir, multichainManifestContent, project);
        }
        else {
            project.manifests.push(projectYamlPath);
        }
    }
    project.root = path_1.default.resolve(project.root);
    // Convert manifest paths to be relative to the project root
    project.manifests = project.manifests.map((manifestPath) => {
        return path_1.default.relative(project.root, manifestPath);
    });
    return project;
}
function addMultichainManifestProjects(parentDir, multichainManifestContent, project) {
    for (const projectPath of multichainManifestContent.projects) {
        const { ext } = path_1.default.parse(projectPath);
        if (!extensionIsYamlOrJSON(ext)) {
            throw new Error(`Extension ${ext} not supported for project ${projectPath}`);
        }
        if (fs_1.default.existsSync(path_1.default.resolve(parentDir, projectPath))) {
            project.manifests.push(path_1.default.resolve(parentDir, projectPath));
        }
        else {
            throw new Error(`Project ${projectPath} not found`);
        }
    }
    if (project.manifests.length === 0) {
        throw new Error('None of the project files specified in the multichain manifest could be found');
    }
}
function getMultichainManifestPath(subquery) {
    const stats = fs_1.default.statSync(subquery);
    let multichainManifestPath;
    let projectRoot;
    if (stats.isDirectory()) {
        projectRoot = subquery;
        const multichainManifestCandidate = path_1.default.resolve(subquery, exports.DEFAULT_MULTICHAIN_MANIFEST);
        if (fs_1.default.existsSync(multichainManifestCandidate)) {
            multichainManifestPath = multichainManifestCandidate;
        }
    }
    else if (stats.isFile()) {
        const { dir, ext } = path_1.default.parse(subquery);
        projectRoot = dir;
        if (extensionIsYamlOrJSON(ext)) {
            const multichainManifestContent = yaml.load(fs_1.default.readFileSync(subquery, 'utf8'));
            if (multichainManifestContent.projects && Array.isArray(multichainManifestContent.projects)) {
                multichainManifestPath = path_1.default.resolve(dir, subquery);
            }
        }
    }
    if (multichainManifestPath && projectRoot) {
        return path_1.default.relative(projectRoot, multichainManifestPath);
    }
    return undefined;
}
function validateSemver(current, required) {
    return (0, semver_1.satisfies)(current, required, { includePrerelease: true });
}
let SemverVersionValidator = class SemverVersionValidator {
    validate(value) {
        if ((0, semver_1.valid)(value) === null) {
            return (0, semver_1.validRange)(value, { includePrerelease: false }) !== null;
        }
        else {
            return (0, semver_1.prerelease)(value || '') === null;
        }
    }
    defaultMessage(args) {
        return `'${args.value}' is not a valid version. Please provide a valid semver`;
    }
};
exports.SemverVersionValidator = SemverVersionValidator;
exports.SemverVersionValidator = SemverVersionValidator = __decorate([
    (0, class_validator_1.ValidatorConstraint)({ name: 'semver', async: false })
], SemverVersionValidator);
async function delay(sec) {
    return new Promise((resolve) => {
        setTimeout(resolve, sec * 1000);
    });
}
function validateObject(object, errorMessage = 'failed to validate object.') {
    const errors = (0, class_validator_1.validateSync)(object, { whitelist: true, forbidNonWhitelisted: true });
    if (errors?.length) {
        const errorMsgs = errors.map((e) => e.toString()).join('\n');
        throw new Error(`${errorMessage}\n${errorMsgs}`);
    }
}
function extensionIsTs(ext) {
    return ext === '.ts';
}
function extensionIsYamlOrJSON(ext) {
    return ext === '.yaml' || ext === '.yml' || ext === '.json';
}
function forbidNonWhitelisted(keys, validationOptions) {
    return function (object, propertyName) {
        (0, class_validator_1.registerDecorator)({
            name: 'forbidNonWhitelisted',
            target: object.constructor,
            propertyName: propertyName,
            constraints: [],
            options: validationOptions,
            validator: {
                validate(value, args) {
                    const isValid = !Object.keys(value).some((key) => !(key in keys));
                    if (!isValid) {
                        throw new Error(`Invalid keys present in value: ${JSON.stringify(value)}. Whitelisted keys: ${JSON.stringify(Object.keys(keys))}`);
                    }
                    return isValid;
                },
            },
        });
    };
}
function IsNetworkEndpoint(cls, validationOptions) {
    return function (object, propertyName) {
        (0, class_validator_1.registerDecorator)({
            name: 'IsNetworkEndpoint',
            target: object.constructor,
            propertyName: propertyName,
            constraints: [],
            options: { message: 'Invalid network endpoint' },
            validator: {
                validate(value, args) {
                    if (typeof value === 'string') {
                        return true;
                    }
                    if (Array.isArray(value)) {
                        return value.every((v) => typeof v === 'string');
                    }
                    if (typeof value === 'object') {
                        return (Object.keys(value).every((v) => typeof v === 'string') &&
                            Object.values(value).every((v) => {
                                const instance = (0, class_transformer_1.plainToInstance)(cls, v);
                                const errors = (0, class_validator_1.validateSync)(instance);
                                return errors === undefined || !errors.length;
                            }));
                    }
                    return false;
                },
            },
        });
    };
}
function normalizeNetworkEndpoints(input, defaultConfig) {
    if (typeof input === 'string') {
        return { [input]: defaultConfig ?? {} };
    }
    else if (Array.isArray(input)) {
        return input.reduce((acc, endpoint) => {
            acc[endpoint] = defaultConfig ?? {};
            return acc;
        }, {});
    }
    for (const key in input) {
        // Yaml can make this null so we ensure it exists
        input[key] = input[key] ?? {};
    }
    return input;
}
function notifyUpdates(pjson, logger) {
    const notifier = (0, update_notifier_1.default)({ pkg: pjson, updateCheckInterval: 0 });
    const latestVersion = notifier.update ? notifier.update.latest : pjson.version;
    if (notifier.update && (0, semver_1.lt)(pjson.version, latestVersion)) {
        logger.info(`Update available: ${pjson.version} → ${latestVersion}`);
    }
    else {
        logger.info(`Current ${pjson.name} version is ${pjson.version}`);
    }
}
function toJsonObject(object) {
    // When using plainToInstance or plainToClass, Map types will need to be converted to a JSON object
    // https://github.com/typestack/class-transformer/issues/1256#issuecomment-1175153352
    return JSON.parse(JSON.stringify(object, (_, value) => {
        if (value instanceof Map) {
            return mapToObject(value);
        }
        return value;
    }));
}
function mapToObject(map) {
    // XXX can use Object.entries with newer versions of node.js
    const assetsObj = {};
    for (const key of map.keys()) {
        assetsObj[key] = map.get(key);
    }
    return assetsObj;
}
let FileReferenceImp = class FileReferenceImp {
    validate(value) {
        if (!value) {
            return false;
        }
        return !!Object.values(value).find((fileReference) => this.isValidFileReference(fileReference));
    }
    defaultMessage(args) {
        return `${JSON.stringify(args.value)} is not a valid assets format`;
    }
    isValidFileReference(fileReference) {
        return (fileReference &&
            typeof fileReference === 'object' &&
            'file' in fileReference &&
            typeof fileReference.file === 'string');
    }
};
exports.FileReferenceImp = FileReferenceImp;
exports.FileReferenceImp = FileReferenceImp = __decorate([
    (0, class_validator_1.ValidatorConstraint)({ name: 'isFileReference', async: false })
], FileReferenceImp);
const tsProjectYamlPath = (tsManifestEntry) => tsManifestEntry.replace('.ts', '.yaml');
exports.tsProjectYamlPath = tsProjectYamlPath;
let IsEndBlockGreater = class IsEndBlockGreater {
    validate(endBlock, args) {
        const object = args.object;
        return object.startBlock !== undefined && object.endBlock !== undefined
            ? object.endBlock >= object.startBlock
            : true;
    }
    defaultMessage(args) {
        return 'End block must be greater than or equal to start block';
    }
};
exports.IsEndBlockGreater = IsEndBlockGreater;
exports.IsEndBlockGreater = IsEndBlockGreater = __decorate([
    (0, class_validator_1.ValidatorConstraint)({ async: false })
], IsEndBlockGreater);
