// @cosmwasm/ts-codegen/main/builder/builder.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TSBuilder = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _wasmAstTypes = require("wasm-ast-types");

var _header = require("../utils/header");

var _path = require("path");

var _fs = require("fs");

var _mkdirp = require("mkdirp");

var _utils = require("../utils");

var _deepmerge = _interopRequireDefault(require("deepmerge"));

var _case = require("case");

var _bundler = require("../bundler");

var _generator = _interopRequireDefault(require("@babel/generator"));

var t = _interopRequireWildcard(require("@babel/types"));

var _reactQuery = require("../plugins/react-query");

var _recoil = require("../plugins/recoil");

var _msgBuilder = require("../plugins/msg-builder");

var _messageComposer = require("../plugins/message-composer");

var _client = require("../plugins/client");

var _types2 = require("../plugins/types");

var _provider = require("../plugins/provider");

var _createHelpers = require("../generators/create-helpers");

var _providerBundle = require("../plugins/provider-bundle");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

var defaultOpts = {
  bundle: {
    enabled: true,
    scope: 'contracts',
    bundleFile: 'bundle.ts'
  }
};
;
;
;
;

function getContract(contractOpt) {
  if (typeof contractOpt === 'string') {
    var name = (0, _path.basename)(contractOpt);
    var contractName = (0, _case.pascal)(name);
    return {
      name: contractName,
      dir: contractOpt
    };
  }

  return {
    name: (0, _case.pascal)(contractOpt.name),
    dir: contractOpt.dir
  };
}

var TSBuilder = /*#__PURE__*/function () {
  function TSBuilder(_ref) {
    var _this = this;

    var contracts = _ref.contracts,
        outPath = _ref.outPath,
        options = _ref.options,
        plugins = _ref.plugins;
    (0, _classCallCheck2["default"])(this, TSBuilder);
    (0, _defineProperty2["default"])(this, "contracts", void 0);
    (0, _defineProperty2["default"])(this, "outPath", void 0);
    (0, _defineProperty2["default"])(this, "options", void 0);
    (0, _defineProperty2["default"])(this, "plugins", []);
    (0, _defineProperty2["default"])(this, "builderContext", new _wasmAstTypes.BuilderContext());
    (0, _defineProperty2["default"])(this, "files", []);
    this.contracts = contracts;
    this.outPath = outPath;
    this.options = (0, _deepmerge["default"])((0, _deepmerge["default"])(_wasmAstTypes.defaultOptions, defaultOpts), options !== null && options !== void 0 ? options : {});
    this.loadDefaultPlugins();

    if (plugins && plugins.length) {
      [].push.apply(this.plugins, plugins);
    }

    this.plugins.forEach(function (plugin) {
      return plugin.setBuilder(_this);
    });
  }

  (0, _createClass2["default"])(TSBuilder, [{
    key: "loadDefaultPlugins",
    value: function loadDefaultPlugins() {
      [].push.apply(this.plugins, [new _types2.TypesPlugin(this.options), new _client.ClientPlugin(this.options), new _messageComposer.MessageComposerPlugin(this.options), new _reactQuery.ReactQueryPlugin(this.options), new _recoil.RecoilPlugin(this.options), new _msgBuilder.MsgBuilderPlugin(this.options), new _provider.ContractsContextProviderPlugin(this.options)]);
    }
  }, {
    key: "build",
    value: function () {
      var _build = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return this.process();

              case 2:
                _context.next = 4;
                return this.after();

              case 4:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function build() {
        return _build.apply(this, arguments);
      }

      return build;
    }() // lifecycle functions

  }, {
    key: "process",
    value: function () {
      var _process = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2() {
        var _iterator, _step, contractOpt, contract, contractInfo;

        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _iterator = _createForOfIteratorHelper(this.contracts);
                _context2.prev = 1;

                _iterator.s();

              case 3:
                if ((_step = _iterator.n()).done) {
                  _context2.next = 13;
                  break;
                }

                contractOpt = _step.value;
                contract = getContract(contractOpt); //resolve contract schema.

                _context2.next = 8;
                return (0, _utils.readSchemas)({
                  schemaDir: contract.dir
                });

              case 8:
                contractInfo = _context2.sent;
                _context2.next = 11;
                return this.render(contract.name, contractInfo);

              case 11:
                _context2.next = 3;
                break;

              case 13:
                _context2.next = 18;
                break;

              case 15:
                _context2.prev = 15;
                _context2.t0 = _context2["catch"](1);

                _iterator.e(_context2.t0);

              case 18:
                _context2.prev = 18;

                _iterator.f();

                return _context2.finish(18);

              case 21:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this, [[1, 15, 18, 21]]);
      }));

      function process() {
        return _process.apply(this, arguments);
      }

      return process;
    }()
  }, {
    key: "render",
    value: function () {
      var _render = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(name, contractInfo) {
        var _iterator2, _step2, plugin, files;

        return _regenerator["default"].wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _iterator2 = _createForOfIteratorHelper(this.plugins);
                _context3.prev = 1;

                _iterator2.s();

              case 3:
                if ((_step2 = _iterator2.n()).done) {
                  _context3.next = 11;
                  break;
                }

                plugin = _step2.value;
                _context3.next = 7;
                return plugin.render(name, contractInfo, this.outPath);

              case 7:
                files = _context3.sent;

                if (files && files.length) {
                  [].push.apply(this.files, files);
                }

              case 9:
                _context3.next = 3;
                break;

              case 11:
                _context3.next = 16;
                break;

              case 13:
                _context3.prev = 13;
                _context3.t0 = _context3["catch"](1);

                _iterator2.e(_context3.t0);

              case 16:
                _context3.prev = 16;

                _iterator2.f();

                return _context3.finish(16);

              case 19:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this, [[1, 13, 16, 19]]);
      }));

      function render(_x, _x2) {
        return _render.apply(this, arguments);
      }

      return render;
    }()
  }, {
    key: "after",
    value: function () {
      var _after = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4() {
        var contractsProviderBundlePlugin, files, helpers;
        return _regenerator["default"].wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                //create useContracts bundle file
                contractsProviderBundlePlugin = new _providerBundle.ContractsProviderBundlePlugin(this.options);
                contractsProviderBundlePlugin.setBuilder(this); //contractContextProviders.ts

                _context4.next = 4;
                return contractsProviderBundlePlugin.render("contractContextProviders", {
                  schemas: []
                }, this.outPath);

              case 4:
                files = _context4.sent;

                if (files && files.length) {
                  [].push.apply(this.files, files);
                }

                helpers = (0, _createHelpers.createHelpers)({
                  outPath: this.outPath,
                  contracts: this.contracts,
                  options: this.options,
                  plugins: this.plugins
                }, this.builderContext);

                if (helpers && helpers.length) {
                  [].push.apply(this.files, helpers);
                }

                if (this.options.bundle.enabled) {
                  this.bundle();
                }

              case 9:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function after() {
        return _after.apply(this, arguments);
      }

      return after;
    }()
  }, {
    key: "bundle",
    value: function () {
      var _bundle = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5() {
        var _this$options$bundle$,
            _this$options,
            _this$options$bundle,
            _this2 = this,
            _this$options2,
            _this$options2$bundle;

        var allFiles, bundleFile, bundlePath, bundleVariables, importPaths, ast, code, _this$options3, _this$options3$bundle;

        return _regenerator["default"].wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                allFiles = this.files;
                bundleFile = this.options.bundle.bundleFile;
                bundlePath = (0, _path.join)((_this$options$bundle$ = (_this$options = this.options) === null || _this$options === void 0 ? void 0 : (_this$options$bundle = _this$options.bundle) === null || _this$options$bundle === void 0 ? void 0 : _this$options$bundle.bundlePath) !== null && _this$options$bundle$ !== void 0 ? _this$options$bundle$ : this.outPath, bundleFile);
                bundleVariables = {};
                importPaths = [];
                allFiles.forEach(function (file) {
                  (0, _bundler.createFileBundle)("".concat(_this2.options.bundle.scope, ".").concat(file.contract), file.filename, bundlePath, importPaths, bundleVariables);
                });
                ast = (0, _bundler.recursiveModuleBundle)(bundleVariables);
                code = (0, _generator["default"])(t.program([].concat(importPaths, (0, _toConsumableArray2["default"])(ast)))).code;

                if ((_this$options2 = this.options) !== null && _this$options2 !== void 0 && (_this$options2$bundle = _this$options2.bundle) !== null && _this$options2$bundle !== void 0 && _this$options2$bundle.bundlePath) {
                  (0, _mkdirp.sync)((_this$options3 = this.options) === null || _this$options3 === void 0 ? void 0 : (_this$options3$bundle = _this$options3.bundle) === null || _this$options3$bundle === void 0 ? void 0 : _this$options3$bundle.bundlePath);
                }

                (0, _mkdirp.sync)(this.outPath);
                if (code.trim() === '') code = 'export {};';
                (0, _fs.writeFileSync)(bundlePath, _header.header + code);

              case 12:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function bundle() {
        return _bundle.apply(this, arguments);
      }

      return bundle;
    }()
  }]);
  return TSBuilder;
}();

exports.TSBuilder = TSBuilder;
