// @subql/common-ethereum/dist/project/types.js
"use strict";
// Copyright 2020-2025 SubQuery Pte Ltd authors & contributors
// SPDX-License-Identifier: GPL-3.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.SubqlEthereumDatasourceKind = exports.SubqlEthereumHandlerKind = exports.EthereumDatasourceKind = exports.EthereumHandlerKind = void 0;
// All of these used to be redefined in this file, re-exporting for simplicity
var types_ethereum_1 = require("@subql/types-ethereum");
Object.defineProperty(exports, "EthereumHandlerKind", { enumerable: true, get: function () { return types_ethereum_1.EthereumHandlerKind; } });
Object.defineProperty(exports, "EthereumDatasourceKind", { enumerable: true, get: function () { return types_ethereum_1.EthereumDatasourceKind; } });
var SubqlEthereumHandlerKind;
(function (SubqlEthereumHandlerKind) {
    SubqlEthereumHandlerKind["EthBlock"] = "ethereum/BlockHandler";
    SubqlEthereumHandlerKind["EthCall"] = "ethereum/TransactionHandler";
    SubqlEthereumHandlerKind["EthEvent"] = "ethereum/LogHandler";
})(SubqlEthereumHandlerKind || (exports.SubqlEthereumHandlerKind = SubqlEthereumHandlerKind = {}));
var SubqlEthereumDatasourceKind;
(function (SubqlEthereumDatasourceKind) {
    SubqlEthereumDatasourceKind["EthRuntime"] = "ethereum/Runtime";
})(SubqlEthereumDatasourceKind || (exports.SubqlEthereumDatasourceKind = SubqlEthereumDatasourceKind = {}));
