// @subql/types-ethereum/dist/project.js
"use strict";
// Copyright 2020-2025 SubQuery Pte Ltd authors & contributors
// SPDX-License-Identifier: GPL-3.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthereumHandlerKind = exports.EthereumDatasourceKind = void 0;
/**
 * Kind of Ethereum datasource.
 * @enum {string}
 */
var EthereumDatasourceKind;
(function (EthereumDatasourceKind) {
    /**
     * The runtime kind of Ethereum datasource.
     */
    EthereumDatasourceKind["Runtime"] = "ethereum/Runtime";
})(EthereumDatasourceKind || (exports.EthereumDatasourceKind = EthereumDatasourceKind = {}));
/**
 * Enum representing the kind of Ethereum handler.
 * @enum {string}
 */
var EthereumHandlerKind;
(function (EthereumHandlerKind) {
    /**
     * Handler for Ethereum blocks.
     */
    EthereumHandlerKind["Block"] = "ethereum/BlockHandler";
    /**
     * Handler for Ethereum transactions.
     */
    EthereumHandlerKind["Call"] = "ethereum/TransactionHandler";
    /**
     * Handler for Ethereum log events.
     */
    EthereumHandlerKind["Event"] = "ethereum/LogHandler";
})(EthereumHandlerKind || (exports.EthereumHandlerKind = EthereumHandlerKind = {}));
