// @cosmwasm/ts-codegen/main/plugins/message-composer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TYPE = exports.MessageComposerPlugin = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _case = require("case");

var w = _interopRequireWildcard(require("wasm-ast-types"));

var _utils = require("../utils");

var _pluginBase = require("./plugin-base");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var TYPE = "message-composer";
exports.TYPE = TYPE;

var MessageComposerPlugin = /*#__PURE__*/function (_BuilderPluginBase) {
  (0, _inherits2["default"])(MessageComposerPlugin, _BuilderPluginBase);

  var _super = _createSuper(MessageComposerPlugin);

  function MessageComposerPlugin() {
    (0, _classCallCheck2["default"])(this, MessageComposerPlugin);
    return _super.apply(this, arguments);
  }

  (0, _createClass2["default"])(MessageComposerPlugin, [{
    key: "initContext",
    value: function initContext(contract, options) {
      return new w.RenderContext(contract, options, this.builder.builderContext);
    }
  }, {
    key: "doRender",
    value: function () {
      var _doRender = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(name, context) {
        var enabled, schemas, localname, TypesFile, ExecuteMsg, typeHash, body, children, TheClass, Interface;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                enabled = this.option.messageComposer.enabled;

                if (enabled) {
                  _context.next = 3;
                  break;
                }

                return _context.abrupt("return");

              case 3:
                schemas = context.contract.schemas;
                localname = (0, _case.pascal)(name) + ".message-composer.ts";
                TypesFile = (0, _case.pascal)(name) + ".types";
                ExecuteMsg = (0, _utils.findExecuteMsg)(schemas);
                _context.next = 9;
                return (0, _utils.findAndParseTypes)(schemas);

              case 9:
                typeHash = _context.sent;
                body = [];
                body.push(w.importStmt(Object.keys(typeHash), "./".concat(TypesFile))); // execute messages

                if (ExecuteMsg) {
                  children = (0, w.getMessageProperties)(ExecuteMsg);

                  if (children.length > 0) {
                    TheClass = (0, _case.pascal)("".concat(name, "MessageComposer"));
                    Interface = (0, _case.pascal)("".concat(name, "Message"));
                    body.push(w.createMessageComposerInterface(context, Interface, ExecuteMsg));
                    body.push(w.createMessageComposerClass(context, TheClass, Interface, ExecuteMsg));
                    context.addProviderInfo(name, w.PROVIDER_TYPES.MESSAGE_COMPOSER_TYPE, TheClass, localname);
                  }
                }

                if (typeHash.hasOwnProperty("Coin")) {
                  // @ts-ignore
                  delete context.utils.Coin;
                }

                return _context.abrupt("return", [{
                  type: TYPE,
                  localname: localname,
                  body: body
                }]);

              case 15:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function doRender(_x, _x2) {
        return _doRender.apply(this, arguments);
      }

      return doRender;
    }()
  }]);
  return MessageComposerPlugin;
}(_pluginBase.BuilderPluginBase);

exports.MessageComposerPlugin = MessageComposerPlugin;
