// @cosmwasm/ts-codegen/main/plugins/provider.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GetLocalNameByContractName = exports.GetLocalBaseNameByContractName = exports.ContractsContextProviderPlugin = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _case = require("case");

var w = _interopRequireWildcard(require("wasm-ast-types"));

var _pluginBase = require("./plugin-base");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var GetLocalNameByContractName = function GetLocalNameByContractName(name) {
  return "".concat((0, _case.pascal)(name), ".provider.ts");
};

exports.GetLocalNameByContractName = GetLocalNameByContractName;

var GetLocalBaseNameByContractName = function GetLocalBaseNameByContractName(name) {
  return "".concat((0, _case.pascal)(name), ".provider");
};

exports.GetLocalBaseNameByContractName = GetLocalBaseNameByContractName;

var ContractsContextProviderPlugin = /*#__PURE__*/function (_BuilderPluginBase) {
  (0, _inherits2["default"])(ContractsContextProviderPlugin, _BuilderPluginBase);

  var _super = _createSuper(ContractsContextProviderPlugin);

  function ContractsContextProviderPlugin(opt) {
    var _this;

    (0, _classCallCheck2["default"])(this, ContractsContextProviderPlugin);
    _this = _super.call(this, opt);
    _this.utils = {
      ContractBase: "__contractContextBase__",
      IContractConstructor: "__contractContextBase__",
      IEmptyClient: "__contractContextBase__"
    };
    return _this;
  }

  (0, _createClass2["default"])(ContractsContextProviderPlugin, [{
    key: "initContext",
    value: function initContext(contract, options) {
      return new w.RenderContext(contract, options, this.builder.builderContext);
    }
  }, {
    key: "doRender",
    value: function () {
      var _doRender = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(name, context) {
        var _this$option, _this$option$useContr, _Object$keys;

        var providerInfo, localname, needEmptyClientType, clientFile, clientClasses, body, signClientProviderInfo, queryClientProviderInfo, messageComposerProviderInfo;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if ((_this$option = this.option) !== null && _this$option !== void 0 && (_this$option$useContr = _this$option.useContractsHooks) !== null && _this$option$useContr !== void 0 && _this$option$useContr.enabled) {
                  _context.next = 2;
                  break;
                }

                return _context.abrupt("return");

              case 2:
                providerInfo = context.getProviderInfos()[name];

                if ((_Object$keys = Object.keys(providerInfo)) !== null && _Object$keys !== void 0 && _Object$keys.length) {
                  _context.next = 5;
                  break;
                }

                return _context.abrupt("return");

              case 5:
                context.addUtil("ContractBase");
                context.addUtil("IContractConstructor");
                localname = GetLocalNameByContractName(name);
                needEmptyClientType = false;
                clientFile = null;
                clientClasses = [];
                body = [];
                signClientProviderInfo = providerInfo[w.PROVIDER_TYPES.SIGNING_CLIENT_TYPE];

                if (signClientProviderInfo) {
                  clientFile = "./".concat(signClientProviderInfo.basename);
                  clientClasses.push(signClientProviderInfo.classname);
                } else {
                  needEmptyClientType = true;
                }

                queryClientProviderInfo = providerInfo[w.PROVIDER_TYPES.QUERY_CLIENT_TYPE];

                if (queryClientProviderInfo) {
                  clientFile = "./".concat(queryClientProviderInfo.basename);
                  clientClasses.push(queryClientProviderInfo.classname);
                } else {
                  needEmptyClientType = true;
                }

                if (clientFile) {
                  body.push(w.importStmt(clientClasses, clientFile));
                }

                messageComposerProviderInfo = providerInfo[w.PROVIDER_TYPES.MESSAGE_COMPOSER_TYPE];

                if (messageComposerProviderInfo) {
                  body.push(w.importStmt([messageComposerProviderInfo.classname], "./".concat(messageComposerProviderInfo.basename)));
                } else {
                  needEmptyClientType = true;
                }

                if (needEmptyClientType) {
                  context.addUtil("IEmptyClient");
                }

                body.push(w.createProvider(name, providerInfo));
                return _context.abrupt("return", [{
                  type: "plugin",
                  pluginType: "provider",
                  localname: localname,
                  body: body
                }]);

              case 22:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function doRender(_x, _x2) {
        return _doRender.apply(this, arguments);
      }

      return doRender;
    }()
  }]);
  return ContractsContextProviderPlugin;
}(_pluginBase.BuilderPluginBase);

exports.ContractsContextProviderPlugin = ContractsContextProviderPlugin;
