// @cosmwasm/ts-codegen/main/plugins/react-query.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ReactQueryPlugin = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _case = require("case");

var w = _interopRequireWildcard(require("wasm-ast-types"));

var _utils = require("../utils");

var _pluginBase = require("./plugin-base");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var ReactQueryPlugin = /*#__PURE__*/function (_BuilderPluginBase) {
  (0, _inherits2["default"])(ReactQueryPlugin, _BuilderPluginBase);

  var _super = _createSuper(ReactQueryPlugin);

  function ReactQueryPlugin() {
    (0, _classCallCheck2["default"])(this, ReactQueryPlugin);
    return _super.apply(this, arguments);
  }

  (0, _createClass2["default"])(ReactQueryPlugin, [{
    key: "initContext",
    value: function initContext(contract, options) {
      return new w.RenderContext(contract, options, this.builder.builderContext);
    }
  }, {
    key: "doRender",
    value: function () {
      var _doRender = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(name, context) {
        var options, enabled, schemas, localname, ContractFile, TypesFile, QueryMsg, ExecuteMsg, typeHash, ExecuteClient, QueryClient, body, clientImports, shouldGenerateMutationHooks;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                options = this.option.reactQuery;
                enabled = options.enabled;

                if (enabled) {
                  _context.next = 4;
                  break;
                }

                return _context.abrupt("return");

              case 4:
                schemas = context.contract.schemas;
                localname = (0, _case.pascal)("".concat(name)) + '.react-query.ts';
                ContractFile = (0, _case.pascal)("".concat(name)) + '.client';
                TypesFile = (0, _case.pascal)("".concat(name)) + '.types';
                QueryMsg = (0, _utils.findQueryMsg)(schemas);
                ExecuteMsg = (0, _utils.findExecuteMsg)(schemas);
                _context.next = 12;
                return (0, _utils.findAndParseTypes)(schemas);

              case 12:
                typeHash = _context.sent;
                ExecuteClient = (0, _case.pascal)("".concat(name, "Client"));
                QueryClient = (0, _case.pascal)("".concat(name, "QueryClient"));
                body = [];
                clientImports = [];
                QueryMsg && clientImports.push(QueryClient); // check that there are commands within the exec msg

                shouldGenerateMutationHooks = ExecuteMsg && (options === null || options === void 0 ? void 0 : options.version) === 'v4' && (options === null || options === void 0 ? void 0 : options.mutations) && (0, w.getMessageProperties)(ExecuteMsg).length > 0;

                if (shouldGenerateMutationHooks) {
                  clientImports.push(ExecuteClient);
                } // general contract imports


                body.push(w.importStmt(Object.keys(typeHash), "./".concat(TypesFile))); // client imports

                body.push(w.importStmt(clientImports, "./".concat(ContractFile))); // query messages

                if (QueryMsg) {
                  [].push.apply(body, w.createReactQueryHooks({
                    context: context,
                    queryMsg: QueryMsg,
                    contractName: name,
                    QueryClient: QueryClient
                  }));
                }

                if (shouldGenerateMutationHooks) {
                  [].push.apply(body, w.createReactQueryMutationHooks({
                    context: context,
                    execMsg: ExecuteMsg,
                    contractName: name,
                    ExecuteClient: ExecuteClient
                  }));
                }

                if (typeHash.hasOwnProperty('Coin')) {
                  // @ts-ignore
                  delete context.utils.Coin;
                }

                return _context.abrupt("return", [{
                  type: 'react-query',
                  localname: localname,
                  body: body
                }]);

              case 26:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function doRender(_x, _x2) {
        return _doRender.apply(this, arguments);
      }

      return doRender;
    }()
  }]);
  return ReactQueryPlugin;
}(_pluginBase.BuilderPluginBase);

exports.ReactQueryPlugin = ReactQueryPlugin;
