// @cosmwasm/ts-codegen/main/plugins/recoil.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RecoilPlugin = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _case = require("case");

var w = _interopRequireWildcard(require("wasm-ast-types"));

var _utils = require("../utils");

var _pluginBase = require("./plugin-base");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var RecoilPlugin = /*#__PURE__*/function (_BuilderPluginBase) {
  (0, _inherits2["default"])(RecoilPlugin, _BuilderPluginBase);

  var _super = _createSuper(RecoilPlugin);

  function RecoilPlugin() {
    var _this;

    (0, _classCallCheck2["default"])(this, RecoilPlugin);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "utils", {
      selectorFamily: 'recoil'
    });
    return _this;
  }

  (0, _createClass2["default"])(RecoilPlugin, [{
    key: "initContext",
    value: function initContext(contract, options) {
      return new w.RenderContext(contract, options, this.builder.builderContext);
    }
  }, {
    key: "doRender",
    value: function () {
      var _doRender = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(name, context) {
        var enabled, schemas, localname, ContractFile, TypesFile, QueryMsg, typeHash, QueryClient, ReadOnlyInstance, body;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                enabled = this.option.recoil.enabled;

                if (enabled) {
                  _context.next = 3;
                  break;
                }

                return _context.abrupt("return");

              case 3:
                schemas = context.contract.schemas;
                localname = (0, _case.pascal)(name) + '.recoil.ts';
                ContractFile = (0, _case.pascal)(name) + '.client';
                TypesFile = (0, _case.pascal)(name) + '.types';
                QueryMsg = (0, _utils.findQueryMsg)(schemas);
                _context.next = 10;
                return (0, _utils.findAndParseTypes)(schemas);

              case 10:
                typeHash = _context.sent;
                QueryClient = null;
                ReadOnlyInstance = null;
                body = [];
                body.push(w.importStmt(['cosmWasmClient'], './chain'));
                body.push(w.importStmt(Object.keys(typeHash), "./".concat(TypesFile))); // query messages

                if (QueryMsg) {
                  QueryClient = (0, _case.pascal)("".concat(name, "QueryClient"));
                  ReadOnlyInstance = (0, _case.pascal)("".concat(name, "ReadOnlyInterface"));
                  body.push(w.importStmt([QueryClient], "./".concat(ContractFile)));
                  body.push(w.createRecoilQueryClientType());
                  body.push(w.createRecoilQueryClient(context, name, QueryClient));
                  [].push.apply(body, w.createRecoilSelectors(context, name, QueryClient, QueryMsg));
                }

                if (typeHash.hasOwnProperty('Coin')) {
                  // @ts-ignore
                  delete context.utils.Coin;
                }

                return _context.abrupt("return", [{
                  type: 'recoil',
                  localname: localname,
                  body: body
                }]);

              case 19:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function doRender(_x, _x2) {
        return _doRender.apply(this, arguments);
      }

      return doRender;
    }()
  }]);
  return RecoilPlugin;
}(_pluginBase.BuilderPluginBase);

exports.RecoilPlugin = RecoilPlugin;
