// @subql/types-near/dist/project.js
"use strict";
// Copyright 2020-2025 SubQuery Pte Ltd authors & contributors
// SPDX-License-Identifier: GPL-3.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.NearHandlerKind = exports.NearDatasourceKind = void 0;
/**
 * Kind of Near datasource.
 * @enum {string}
 */
var NearDatasourceKind;
(function (NearDatasourceKind) {
    /**
     * The runtime kind of Near datasource.
     */
    NearDatasourceKind["Runtime"] = "near/Runtime";
})(NearDatasourceKind || (exports.NearDatasourceKind = NearDatasourceKind = {}));
/**
 * Enum representing the kind of Near handler.
 * @enum {string}
 */
var NearHandlerKind;
(function (NearHandlerKind) {
    /**
     * Handler for Near blocks.
     */
    NearHandlerKind["Block"] = "near/BlockHandler";
    /**
     * Handler for Near transactions.
     */
    NearHandlerKind["Transaction"] = "near/TransactionHandler";
    /**
     * Handler for Near actions.
     */
    NearHandlerKind["Action"] = "near/ActionHandler";
    /**
     * Handler for Near transactions receipts.
     */
    NearHandlerKind["Receipt"] = "near/ReceiptHandler";
})(NearHandlerKind || (exports.NearHandlerKind = NearHandlerKind = {}));
