// @subql/types-cosmos/dist/project.js
"use strict";
// // Copyright 2020-2025 SubQuery Pte Ltd authors & contributors
// SPDX-License-Identifier: GPL-3.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.CosmosHandlerKind = exports.CosmosDatasourceKind = void 0;
var CosmosDatasourceKind;
(function (CosmosDatasourceKind) {
    CosmosDatasourceKind["Runtime"] = "cosmos/Runtime";
    CosmosDatasourceKind["Custom"] = "cosmos/Custom";
})(CosmosDatasourceKind || (exports.CosmosDatasourceKind = CosmosDatasourceKind = {}));
/**
 * Enum representing the kind of Cosmos handler.
 * @enum {string}
 */
var CosmosHandlerKind;
(function (CosmosHandlerKind) {
    /**
     * Handler for Cosmos blocks.
     */
    CosmosHandlerKind["Block"] = "cosmos/BlockHandler";
    /**
     * Handler for Cosmos transactions.
     */
    CosmosHandlerKind["Transaction"] = "cosmos/TransactionHandler";
    /**
     * Handler for Cosmos messages.
     */
    CosmosHandlerKind["Message"] = "cosmos/MessageHandler";
    /**
     * Handler for Cosmos events.
     */
    CosmosHandlerKind["Event"] = "cosmos/EventHandler";
})(CosmosHandlerKind || (exports.CosmosHandlerKind = CosmosHandlerKind = {}));
