// @subql/common-ethereum/dist/project/versioned/v1_0_0/model.js
"use strict";
// Copyright 2020-2025 SubQuery Pte Ltd authors & contributors
// SPDX-License-Identifier: GPL-3.0
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectManifestV1_0_0Impl = exports.DeploymentV1_0_0 = exports.ProjectNetworkV1_0_0 = exports.EthereumEndpointConfig = exports.ProjectNetworkDeploymentV1_0_0 = exports.EthereumRunnerSpecsImpl = exports.CustomDatasourceTemplateImpl = exports.RuntimeDatasourceTemplateImpl = exports.EthereumCustomDataSourceImpl = exports.EthereumRuntimeDataSourceImpl = exports.EthereumRunnerNodeImpl = void 0;
const common_1 = require("@subql/common");
const class_transformer_1 = require("class-transformer");
const class_validator_1 = require("class-validator");
const models_1 = require("../../models");
const Ethereum_NODE_NAME = `@subql/node-ethereum`;
class EthereumRunnerNodeImpl extends common_1.RunnerNodeImpl {
    name = Ethereum_NODE_NAME;
}
exports.EthereumRunnerNodeImpl = EthereumRunnerNodeImpl;
__decorate([
    (0, class_validator_1.IsIn)([Ethereum_NODE_NAME], {
        message: `Runner Substrate node name incorrect, suppose be '${Ethereum_NODE_NAME}'`,
    }),
    __metadata("design:type", String)
], EthereumRunnerNodeImpl.prototype, "name", void 0);
function validateObject(object, errorMessage = 'failed to validate object.') {
    const errors = (0, class_validator_1.validateSync)(object, { whitelist: true, forbidNonWhitelisted: true });
    if (errors?.length) {
        const errorMsgs = errors.map((e) => e.toString()).join('\n');
        throw new Error(`${errorMessage}\n${errorMsgs}`);
    }
}
class EthereumRuntimeDataSourceImpl extends models_1.RuntimeDataSourceBase {
    validate() {
        return validateObject(this, 'failed to validate runtime datasource.');
    }
}
exports.EthereumRuntimeDataSourceImpl = EthereumRuntimeDataSourceImpl;
class EthereumCustomDataSourceImpl extends models_1.CustomDataSourceBase {
    validate() {
        return validateObject(this, 'failed to validate custom datasource.');
    }
}
exports.EthereumCustomDataSourceImpl = EthereumCustomDataSourceImpl;
class RuntimeDatasourceTemplateImpl extends EthereumRuntimeDataSourceImpl {
    name;
}
exports.RuntimeDatasourceTemplateImpl = RuntimeDatasourceTemplateImpl;
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], RuntimeDatasourceTemplateImpl.prototype, "name", void 0);
class CustomDatasourceTemplateImpl extends EthereumCustomDataSourceImpl {
    name;
}
exports.CustomDatasourceTemplateImpl = CustomDatasourceTemplateImpl;
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], CustomDatasourceTemplateImpl.prototype, "name", void 0);
class EthereumRunnerSpecsImpl {
    node;
    query;
}
exports.EthereumRunnerSpecsImpl = EthereumRunnerSpecsImpl;
__decorate([
    (0, class_validator_1.IsObject)(),
    (0, class_validator_1.ValidateNested)(),
    (0, class_transformer_1.Type)(() => EthereumRunnerNodeImpl),
    __metadata("design:type", Object)
], EthereumRunnerSpecsImpl.prototype, "node", void 0);
__decorate([
    (0, class_validator_1.IsObject)(),
    (0, class_validator_1.ValidateNested)(),
    (0, class_transformer_1.Type)(() => common_1.RunnerQueryBaseModel),
    __metadata("design:type", Object)
], EthereumRunnerSpecsImpl.prototype, "query", void 0);
class ProjectNetworkDeploymentV1_0_0 {
    chainId;
    bypassBlocks;
}
exports.ProjectNetworkDeploymentV1_0_0 = ProjectNetworkDeploymentV1_0_0;
__decorate([
    (0, class_validator_1.IsNotEmpty)(),
    (0, class_transformer_1.Transform)(({ value }) => value.trim()),
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], ProjectNetworkDeploymentV1_0_0.prototype, "chainId", void 0);
__decorate([
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.IsArray)(),
    __metadata("design:type", Array)
], ProjectNetworkDeploymentV1_0_0.prototype, "bypassBlocks", void 0);
class EthereumEndpointConfig extends common_1.CommonEndpointConfig {
    batchSize;
}
exports.EthereumEndpointConfig = EthereumEndpointConfig;
__decorate([
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.IsPositive)(),
    __metadata("design:type", Number)
], EthereumEndpointConfig.prototype, "batchSize", void 0);
class ProjectNetworkV1_0_0 extends common_1.CommonProjectNetworkV1_0_0 {
    endpoint = {};
}
exports.ProjectNetworkV1_0_0 = ProjectNetworkV1_0_0;
__decorate([
    (0, class_validator_1.IsOptional)(),
    (0, common_1.IsNetworkEndpoint)(EthereumEndpointConfig),
    __metadata("design:type", Object)
], ProjectNetworkV1_0_0.prototype, "endpoint", void 0);
class DeploymentV1_0_0 extends common_1.BaseDeploymentV1_0_0 {
    network;
    runner;
    dataSources;
    templates;
}
exports.DeploymentV1_0_0 = DeploymentV1_0_0;
__decorate([
    (0, class_transformer_1.Transform)((params) => {
        if (params.value.genesisHash && !params.value.chainId) {
            params.value.chainId = params.value.genesisHash;
        }
        return (0, class_transformer_1.plainToInstance)(ProjectNetworkDeploymentV1_0_0, params.value);
    }),
    (0, class_validator_1.ValidateNested)(),
    (0, class_transformer_1.Type)(() => ProjectNetworkDeploymentV1_0_0),
    __metadata("design:type", ProjectNetworkDeploymentV1_0_0)
], DeploymentV1_0_0.prototype, "network", void 0);
__decorate([
    (0, class_validator_1.IsObject)(),
    (0, class_validator_1.ValidateNested)(),
    (0, class_transformer_1.Type)(() => EthereumRunnerSpecsImpl),
    __metadata("design:type", Object)
], DeploymentV1_0_0.prototype, "runner", void 0);
__decorate([
    (0, class_validator_1.IsArray)(),
    (0, class_validator_1.ValidateNested)(),
    (0, class_transformer_1.Type)(() => EthereumCustomDataSourceImpl, {
        discriminator: {
            property: 'kind',
            subTypes: [{ value: EthereumRuntimeDataSourceImpl, name: 'ethereum/Runtime' }],
        },
        keepDiscriminatorProperty: true,
    }),
    __metadata("design:type", Array)
], DeploymentV1_0_0.prototype, "dataSources", void 0);
__decorate([
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.IsArray)(),
    (0, class_validator_1.ValidateNested)(),
    (0, class_transformer_1.Type)(() => CustomDatasourceTemplateImpl, {
        discriminator: {
            property: 'kind',
            subTypes: [{ value: RuntimeDatasourceTemplateImpl, name: 'ethereum/Runtime' }],
        },
        keepDiscriminatorProperty: true,
    }),
    __metadata("design:type", Array)
], DeploymentV1_0_0.prototype, "templates", void 0);
class ProjectManifestV1_0_0Impl extends common_1.ProjectManifestBaseImpl {
    constructor() {
        super(DeploymentV1_0_0);
    }
    specVersion = '1.0.0';
    dataSources;
    network;
    name;
    version;
    schema;
    templates;
    runner;
    parent;
}
exports.ProjectManifestV1_0_0Impl = ProjectManifestV1_0_0Impl;
__decorate([
    (0, class_validator_1.Equals)('1.0.0'),
    __metadata("design:type", Object)
], ProjectManifestV1_0_0Impl.prototype, "specVersion", void 0);
__decorate([
    (0, class_transformer_1.Type)(() => EthereumCustomDataSourceImpl, {
        discriminator: {
            property: 'kind',
            subTypes: [{ value: EthereumRuntimeDataSourceImpl, name: 'ethereum/Runtime' }],
        },
        keepDiscriminatorProperty: true,
    }),
    __metadata("design:type", Array)
], ProjectManifestV1_0_0Impl.prototype, "dataSources", void 0);
__decorate([
    (0, class_transformer_1.Type)(() => ProjectNetworkV1_0_0),
    __metadata("design:type", ProjectNetworkV1_0_0)
], ProjectManifestV1_0_0Impl.prototype, "network", void 0);
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], ProjectManifestV1_0_0Impl.prototype, "name", void 0);
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], ProjectManifestV1_0_0Impl.prototype, "version", void 0);
__decorate([
    (0, class_validator_1.ValidateNested)(),
    (0, class_transformer_1.Type)(() => common_1.FileType),
    __metadata("design:type", common_1.FileType)
], ProjectManifestV1_0_0Impl.prototype, "schema", void 0);
__decorate([
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.IsArray)(),
    (0, class_validator_1.ValidateNested)(),
    (0, class_transformer_1.Type)(() => CustomDatasourceTemplateImpl, {
        discriminator: {
            property: 'kind',
            subTypes: [{ value: RuntimeDatasourceTemplateImpl, name: 'ethereum/Runtime' }],
        },
        keepDiscriminatorProperty: true,
    }),
    __metadata("design:type", Array)
], ProjectManifestV1_0_0Impl.prototype, "templates", void 0);
__decorate([
    (0, class_validator_1.IsObject)(),
    (0, class_validator_1.ValidateNested)(),
    (0, class_transformer_1.Type)(() => EthereumRunnerSpecsImpl),
    __metadata("design:type", Object)
], ProjectManifestV1_0_0Impl.prototype, "runner", void 0);
__decorate([
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.IsObject)(),
    (0, class_transformer_1.Type)(() => common_1.ParentProjectModel),
    __metadata("design:type", Object)
], ProjectManifestV1_0_0Impl.prototype, "parent", void 0);
