// @subql/common-near/dist/project/utils.js
"use strict";
// Copyright 2020-2025 SubQuery Pte Ltd authors & contributors
// SPDX-License-Identifier: GPL-3.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.isBlockHandlerProcessor = isBlockHandlerProcessor;
exports.isTransactionHandlerProcessor = isTransactionHandlerProcessor;
exports.isActionHandlerProcessor = isActionHandlerProcessor;
exports.isReceiptHandlerProcessor = isReceiptHandlerProcessor;
exports.isCustomDs = isCustomDs;
exports.isRuntimeDs = isRuntimeDs;
const types_near_1 = require("@subql/types-near");
function isBlockHandlerProcessor(hp) {
    return hp.baseHandlerKind === types_near_1.NearHandlerKind.Block;
}
function isTransactionHandlerProcessor(hp) {
    return hp.baseHandlerKind === types_near_1.NearHandlerKind.Transaction;
}
function isActionHandlerProcessor(hp) {
    return hp.baseHandlerKind === types_near_1.NearHandlerKind.Action;
}
function isReceiptHandlerProcessor(hp) {
    return hp.baseHandlerKind === types_near_1.NearHandlerKind.Receipt;
}
function isCustomDs(ds) {
    return ds.kind !== types_near_1.NearDatasourceKind.Runtime && !!ds.processor;
}
function isRuntimeDs(ds) {
    return ds.kind === types_near_1.NearDatasourceKind.Runtime;
}
